src/network/tcp/TcpSocket.hhi

```diff
@@ -0,0 +1,127 @@
+<?hh // strict
+
+namespace HH\Lib\Experimental\Network;
+
+use namespace HH\Lib\Experimental\IO;
+use type Throwable;
+
+/**
+ * TCP socket connection.
+ */
+final class TcpSocket implements SocketHandle {
+  /**
+   * Disables Nagle's Algorithm when set.
+   */
+  const int NODELAY = 100;
+
+  /**
+   * Sets the TCP keep-alive timeout in seconds, 0 to disable keep-alive.
+   */
+  const int KEEPALIVE = 101;
+
+  /**
+   * Sockets are created using connect() or TcpServer::accept().
+   */
+  private function __construct();
+
+  /**
+   * Connect to the given peer (will automatically perform a DNS lookup for host names).
+   */
+  public static function connect(
+    string $host,
+    int $port,
+    ?TlsClientEncryption $tls = null,
+  ): Awaitable<TcpSocket>;
+
+  /**
+   * Returns a pair of connected TCP sockets.
```

Comment:
What native call would this be backed by in an ideal world, and when would it be desired instead of `HH\IO\pipe()` ?

Reply:
`HH\Lib\IO\pipe` doesn't return `TcpSocket` instance ?

Reply:
To rephrase: in what circumstances are TcpSocket pipes needed instead of native pipes?

Reply:
yes. socket resource can be used with `socket_` functions, the `HH\Lib\Network\Socket` provieds more functionality ( get the socket address and port, set socket options )
`TcpSocket` provides ability to get the remote address and port.

Reply:
> socket resource can be used with `socket_` functions

not releveant for the HSL

> provieds more functionality ( get the socket address and port, set socket options )
`TcpSocket` provides ability to get the remote address and port.

Why is this useful if you got them from a pipe? remote address will always be localhost, and port will be something random, connected to the other one

Reply:
if you don't see it being useful i can just remove it, but i think we should keep it as a replacement for `socket_create_pair` builtin 
see : https://secure.php.net/manual/en/function.socket-create-pair.php